R/cube-comparisons.R

```diff
@@ -0,0 +1,306 @@
+compareDimsPairwise <- function(cube, dim = c("cols", "rows"), baseline, value=c("statistic", "p.value")) {
+    dim <- match.arg(dim)
+    
+    # grab the names of extents for each dimensions
+    # TODO: we shouldn't need to do this if we can just pass baseline/x into the
+    # subsetting function when we can subset by name _or_ id natively
+    if (dim == "cols") {
+        names <- colnames(as.array(cube))
+    } else if (dim == "rows") {
+        names <- rownames(as.array(cube))
+    }
+    
+    to_compare <- names[!(names %in% baseline)]
+    
+    out <- vapply(
+        names, function(one_extent) {
+            # generate the 2xm or nx2 table for testing
+            return(compareDims(cube, baseline = baseline, x = one_extent, dim = dim, value=value))
+        }
+        , numeric(1)
+    )
+    
+    out <- simplify2array(out)
+    
+    # if the dimension is rows, the array that vapply returns needs to be
+    # transposed
+    if (dim == "rows") {
+        out <- t(out)
+    }
+    
+    names(out) <- names
+    
+    return(out)
+}
+
+#' Matrix of Chi-Squared Statistics for all rows or columns
+#' 
+#' Generate a matrix of pairwise comparisons of rows or columns, each against 
+#' the others.
+#'
+#' @param cube a CrunchCube representing counts of a categorical by categorical
+#' contingency table.
+#' @param dim which dimension along which to compare (`cols` or `rows`)
+#' @param value Value of the test to return. Defaults to `statistic`; use `p.value`
+#' for the p-value of the test of the pair of columns/rows.
+#'
+#' @return A symmetric square matrix of all column-comparison or row-comparison
+#' \eqn{\chi^2} statistics. Typical element \eqn{i,j} is the test statistic equivalent
+#' to \code{chisq.test} subsetting to just columns i and j (for dim="cols").
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' some_cube <- crunch_example_data(cat_by_cat)
+#' 
+#' # By default, return the 'statistic' of chisq.test, evaluated
+#' # for each pair of columns pairwise:
+#' pairwiseMatrix(some_cube)
+#' 
+#' # or each pair of rows:
+#' pairwiseMatrix(some_cube, dim="rows")
+#' 
+#' # Or, specify another "value" from `chisq.test`:
+#' pairwiseMatrix(some_cube, value="p.value")
+#' }
+pairwiseMatrix <- function(cube, dim=c("cols", "rows"), value=c('statistic', 'p.value')){
+    dim <- match.arg(dim)
+    value <- match.arg(value)
+    a <- as.array(cube)
+    if (dim == "cols") {
+        names <- colnames(a)
+    } else if (dim == "rows") {
+        names <- rownames(a)
+    }
+    out <- vapply(names, function(cur) {
+            compareDimsPairwise(cube, dim=dim, baseline=cur, value=value)
+        },
+        numeric(length(names))
+    )
+    return(simplify2array(out))
+}
+
+#' Matrix of Chi-Squared Statistics for all rows or columns
+#' 
+#' Use the alternative Wishart method of forming the matrix of column- or row-wise
+#' comparison Chi-squared test statistics for a categorical-by-categorical 
+#' contingency table.
+#'
```

Comment:
To add:

The null hypothesis is that all of the rows (or columns) are equal to each other. The test statistic matrix that is returned when requested is a measure of closeness between the pair of rows (or columns) given by their names. The p-value matrix that is returned are similarly the probabilities of finding the observed or more extreme results while the null hypothesis is true for each pair of rows (or columns).